signer/core/signed_data.go

```diff
@@ -0,0 +1,633 @@
+	// Get dependencies primary first, then alphabetical
+	deps := typedData.Dependencies(primaryType, []string{})
+	for i, dep := range deps {
+		if dep == primaryType {
+			deps = append(deps[:i], deps[i+1:]...)
+			break
+		}
+	}
+	deps = append([]string{primaryType}, deps...)
+
+	// Format as a string with fields
+	var buffer bytes.Buffer
+	for _, dep := range deps {
+		buffer.WriteString(dep)
+		buffer.WriteString("(")
+		for _, obj := range typedData.Types[dep] {
+			buffer.WriteString(obj["type"])
+			buffer.WriteString(" ")
+			buffer.WriteString(obj["name"])
+			buffer.WriteString(",")
+		}
+		buffer.Truncate(buffer.Len() - 1)
+		buffer.WriteString(")")
+	}
+	return buffer.Bytes()
+}
+
+func (typedData *TypedData) TypeHash(primaryType string) hexutil.Bytes {
+	return crypto.Keccak256(typedData.EncodeType(primaryType))
+}
+
+// EncodeData generates the following encoding:
+// `enc(value₁) ‖ enc(value₂) ‖ … ‖ enc(valueₙ)`
+//
+// each encoded member is 32-byte long
+func (typedData *TypedData) EncodeData(primaryType string, data map[string]interface{}) hexutil.Bytes {
+	encTypes := []string{}
+	encValues := []interface{}{}
+
+	// Add typehash
+	encTypes = append(encTypes, "bytes32")
+	encValues = append(encValues, typedData.TypeHash(primaryType))
+
+	// Handle primitive values
+	handlePrimitiveValue := func(encType string, encValue interface{}) (string, interface{}) {
+		var primitiveEncType string
+		var primitiveEncValue interface{}
+
+		switch encType {
+		case "address":
+			primitiveEncType = "uint160"
+			bytesValue := hexutil.Bytes{}
+			for i := 0; i < 12; i++ {
+				bytesValue = append(bytesValue, 0)
+			}
+			for _, _byte := range common.HexToAddress(encValue.(string)) {
+				bytesValue = append(bytesValue, _byte)
+			}
+			primitiveEncValue = bytesValue
+		case "bool":
+			primitiveEncType = "uint256"
+			var int64Val int64
+			if encValue.(bool) {
+				int64Val = 1
+			}
+			primitiveEncValue = abi.U256(big.NewInt(int64Val))
+		case "bytes", "string":
+			primitiveEncType = "bytes32"
+			primitiveEncValue = crypto.Keccak256(bytesValueOf(encValue))
+		default:
+			if strings.HasPrefix(encType, "bytes") {
+				encTypes = append(encTypes, "bytes32")
+				sizeStr := strings.TrimPrefix(encType, "bytes")
+				size, _ := strconv.Atoi(sizeStr)
+				bytesValue := hexutil.Bytes{}
+				for i := 0; i < 32-size; i++ {
+					bytesValue = append(bytesValue, 0)
+				}
+				bytesValue = append(bytesValue, encValue.(hexutil.Bytes)...)
+				primitiveEncValue = bytesValue
+			} else if strings.HasPrefix(encType, "uint") || strings.HasPrefix(encType, "int") {
+				primitiveEncType = "uint256"
+				primitiveEncValue = abi.U256(encValue.(*big.Int))
+			}
+		}
+		return primitiveEncType, primitiveEncValue
+	}
+
+	// Add field contents. Structs and arrays have special handlings.
+	for _, field := range typedData.Types[primaryType] {
+		encType := field["type"]
+		encValue := data[field["name"]]
+		if encType[len(encType)-1:] == "]" {
```

Comment:
Arrays are not well specified yet in EIP712. The approach here is to encode an array as a hash of the elements, which makes sense. This may be handled differently for fixed and dynamic sized arrays, jury is till out on that one.

Reply:
Should we just disallow dynamically sized arrays for now? That is, throw an error when the user doesn't specify the size.

Reply:
Sounds better to abort signing than signing something not well specified, at this stage